### PR TITLE
detect/proto: support gre protocol in rule selection

### PR DESCRIPTION
**Pull request by thomasjwinter**
This allows the GRE rules to be specific to GRE decoded traffic and not checked against non GRE packets.

Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://docs.suricata.io/en/latest/devguide/codebase/contributing/contribution-process.html
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/6261

Describe changes:
-
-
-

### Provide values to any of the below to override the defaults.

To use a pull request use a branch name like `pr/N` where `N` is the
pull request number.

Alternatively, `SV_BRANCH` may also be a link to an
OISF/suricata-verify pull-request.

```
SV_REPO=
SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1350
SU_REPO=
SU_BRANCH=
LIBHTP_REPO=
LIBHTP_BRANCH=
```

### CODE REVIEW

**Comment by victorjulien** [comment]
I'm ok with adding this, but we'll need:
- a ticket
- an update to the user guide
- Suricata-Verify tests to show it works

Btw, I believe that `alert ip ... ip_proto:47; ... ` would have the same effect.

**Comment by thomasjwinter** [comment]
https://github.com/OISF/suricata-verify/pull/1350

**Review by jufajardini — COMMENTED** [review]
Added the current SV pr number and re-triggered related CI checks. Not sure if we'd like to see something more for documentation.

Have reviewed SV PR.

File: doc/userguide/rules/intro.rst [hunk 1]
```diff
@@ -69,12 +69,18 @@ Protocol
...
 * tcp (for tcp-traffic)
 * udp
-* icmp
-* ip (ip stands for 'all' or 'any')
+* icmpv4
+* icmpv6
+* icmp (both icmpv4 and icmpv6)
+* sctp
+* gre
+* ip4 or ipv4 ('all' or 'any' IPv4 traffic)
```
Comment (jasonish) [inline doc/userguide/rules/intro.rst:81]: I know this was in the existing code, but `'all' or 'any'` is a bit misleading as it makes them look like literals that can be used as an alias.  This should probably be:
```
* ip4 or ipv4
* ip6 or ipv6
* ip (both ipv4 and ipv6)
```
The first 2 lines probably don't need the extra annotation as they are as clear as `udp` or `icmp`.

**Comment by catenacyber** [comment]
Thanks Thomas

But as GRE is some tunneling protocol, we can have UDP over GRE
How do I say that I want a rule only for this kind of traffic ? (UDP over GRE)
Could there be something more generic for all tunneling protocols ?

**Comment by jufajardini** [comment]
A bit of a related ticket: https://redmine.openinfosecfoundation.org/issues/5660

**Review by catenacyber — CHANGES_REQUESTED** [review]
@thomasjwinter are you planning to address the comments ?

**Comment by thomasjwinter** [comment]
Probably not any time soon, sorry. I don't know the answer to differentiating various tunneled headers.

We've had this patch applied to our suricata stream since 2016. Previously had gre decoder events using pkthdr then changed to gre with this patch. Why it wasn't simply changed to "ip_proto:47" I can't say.

**Comment by victorjulien** [comment]
Ok, thanks. Closing for now. We can continue to consider these things in the ticket.

**Pull request closed without merging**